### PR TITLE
IInvocationExpression and IArgument refinement and testing

### PR DESCRIPTION
**Pull request by msJohnHamby**
This change set removes syntactic concerns from IInvocationExpression and IArgument. Specifically, the ArgumentsInSourceOrder property of IInvocationExpression has been replaced with ArgumentsInEvaluationOrder, and the ArgumentKind property of IArgument has been removed.

Those interested only in the API surface are can confine their attention to changes to the IExpression.cs file.

This change set also includes new tests to cover the surface areas of these interfaces for a variety of circumstances, along with a set of bug fixes to make the tests pass.

This change set does not includes fixes for all previously-reported issues related to invocations. For example, there are still some problems with matching array arguments to paramarray parameters, and the reported issues remain open for the time being.

@dotnet/roslyn-compiler @dotnet/roslyn-analysis @AlekseyTs @jcouv @genlu @jaredpar @DustinCampbell @nguerrera @Srivatsn @gafter @mattwar @CyrusNajmabadi @mavasani @ManishJayaswal  and anyone else eager to read 1600 lines of new tests are requested to review.

### CODE REVIEW

File: src/Compilers/Core/Portable/Compilation/IExpression.cs [hunk 1]
```diff
@@ -46,11 +46,11 @@ public interface IInvocationExpression : IHasArgumentsExpression
...
         /// <summary>
-        /// Arguments of the invocation, excluding the instance argument. Arguments are in the order specified in source,
-        /// and params/ParamArray arguments have been collected into arrays. Arguments are not present
-        /// unless supplied in source.
+        /// Arguments of the invocation, excluding the instance argument. Arguments are in the order that they will be evaluated,
+        /// and params/ParamArray arguments have been collected into arrays. Default values are supplied for
+        /// optional arguments missing in source.
         /// </summary>
-        ImmutableArray<IArgument> ArgumentsInSourceOrder { get; }
+        ImmutableArray<IArgument> ArgumentsInEvaluationOrder { get; }
```
Comment (nguerrera) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: Why isn't this on IHasArgumentsExpression?

**Comment by nguerrera** [comment]
API change looks good. (I didn't review the tests.)

Now that you have the hood up on argument handling, it might be a good time to look into vararg (`__arglist`) handling. See #8394.

File: src/Compilers/Core/Portable/Compilation/IExpression.cs (hunk 1 above)
Comment (JohnHamby) [inline src/Compilers/Core/Portable/Compilation/IExpression.cs]: It should be.

**Comment by JohnHamby** [comment]
I have pushed a commit that moves ArgumentsInEvaluationOrder from IInvocationExpresion to IHasArgumentsExpression. Tests for the additional cases in other interfaces derived from IHasArgumentsExpression will be added later in the test wave when those interfaces receive exhaustive testing.

**Comment by JohnHamby** [comment]
retest windows_debug_unit32_prtest please

**Comment by JohnHamby** [comment]
retest windows_debug_unit64_prtest please

**Comment by JohnHamby** [comment]
retest windows_debug_unit32_prtest please
retest windows_debug_unit64_prtest please
retest windows_eta_open_prtest please

**Comment by JohnHamby** [comment]
Adding @VSadov @cston in hopes of attracting compiler team members.

**Comment by genlu** [comment]
>         private abstract class ArgumentBase : IArgument

I think `ArgumentBase` is not used anywhere else, maybe merge `Argument` and `ArgumentBase` into one concrete type instead.

---
Refers to: src/Compilers/CSharp/Portable/BoundTree/Expression.cs:240 in 4e5617e. [](commit_id = 4e5617efc1397a52b72d356129d31bb9a9038567, deletion_comment = False)

**Comment by JohnHamby** [comment]
I've pushed a new commit that removes the ArgumentBase class.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 2]
```diff
@@ -83,7 +67,7 @@ public override void Accept(OperationVisitor visitor)
             return visitor.VisitInvocationExpression(this, argument);
         }
 
-        internal static ImmutableArray<IArgument> DeriveArguments(ImmutableArray<BoundExpression> boundArguments, ImmutableArray<string> argumentNames, ImmutableArray<int> argumentsToParameters, ImmutableArray<RefKind> argumentRefKinds, ImmutableArray<Symbols.ParameterSymbol> parameters, SyntaxNode invocationSyntax)
+        internal static ImmutableArray<IArgument> DeriveArgumentsInParameterOrder(ImmutableArray<BoundExpression> boundArguments, ImmutableArray<string> argumentNames, ImmutableArray<int> argumentsToParameters, ImmutableArray<RefKind> argumentRefKinds, ImmutableArray<Symbols.ParameterSymbol> parameters, SyntaxNode invocationSyntax)
```
Comment (genlu) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Consider move `DeriveArgumentsInParameterOrder` and `DeriveArgumentsInEvaluationOrder` both to `Expression` class at the end of this file, since they are called by indexer and object creation as well.

**Comment by genlu** [comment]
>             // base.VisitInvocationExpression only visit operations in ArgumentsInSourceOrder

I think I left the wrong comment here. OperationWalker.VisitInvocationExpression actually visits operations from `ArgumentsInParameterOrder`. So could you fix this for me as well and visit `ArgumentInEvaluationOrder` here in this method?
(Or vice versa, visit `ArgumentInEvaluationOrder` in OperationWalker and verify `ArgumentsInParameterOrder` here, which might make more sense from a semantic point of view)

---
Refers to: src/Test/Utilities/Shared/Compilation/TestOperationWalker.cs:207 in 4e5617e. [](commit_id = 4e5617efc1397a52b72d356129d31bb9a9038567, deletion_comment = False)

**Comment by genlu** [comment]
>         {

Need to add checks for `ArgumentInEvaluationOrder` in `VisitObjectCreationExpression` and `VisitIndexedPropertyReferenceExpression` as well. (Or change walker to traverse `ArgumentInEvaluationOrder` instead and verify `ArgumentsInParameterOrder` in the test walker)

---
Refers to: src/Test/Utilities/Shared/Compilation/TestOperationWalker.cs:430 in 4e5617e. [](commit_id = 4e5617efc1397a52b72d356129d31bb9a9038567, deletion_comment = False)

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 3]
```diff
@@ -297,9 +297,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return ArgumentMatchingParameter(Me.Arguments, parameter, Me.Method.Parameters)
         End Function
 
-        Private ReadOnly Property IInvocationExpression_ArgumentsInSourceOrder As ImmutableArray(Of IArgument) Implements IInvocationExpression.ArgumentsInSourceOrder
+        Private ReadOnly Property IHasArgumentsExpression_ArgumentsInSourceOrder As ImmutableArray(Of IArgument) Implements IHasArgumentsExpression.ArgumentsInEvaluationOrder
```
Comment (genlu) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: property name should be `IHasArgumentsExpression_ArgumentsInEvaluationOrder`

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 4]
```diff
@@ -312,9 +312,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         Private ReadOnly Property IInvocationExpression_IsVirtual As Boolean Implements IInvocationExpression.IsVirtual
```
Comment (genlu) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: The code is the same as in `IInvocationExpression_IsVirtual` above, consider create a helper.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 5]
```diff
@@ -1533,7 +1539,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
         Private ReadOnly Property IMethodBindingExpression_IsVirtual As Boolean Implements IMethodBindingExpression.IsVirtual
             Get
-                Return Me.Method IsNot Nothing AndAlso (Me.Method.IsOverridable OrElse Me.Method.IsOverrides OrElse Me.Method.IsMustOverride) AndAlso Not Me.SuppressVirtualCalls
+                Dim method As IMethodSymbol = Me.Method
+                Return method IsNot Nothing AndAlso Me.ReceiverOpt IsNot Nothing AndAlso (method.IsVirtual OrElse method.IsAbstract OrElse method.IsOverride) AndAlso Not Me.ReceiverOpt.SuppressVirtualCalls
             End Get
```
Comment (genlu) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: The code is the same as in `IInvocationExpression_IsVirtual` above, consider create a helper.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 6]
```diff
@@ -493,22 +482,27 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
-                    Return ArgumentKind.Positional
-                End Get
-            End Property
-
             Public Overrides ReadOnly Property InConversion As IOperation
                 Get
-                    Return _argument.InConversion
+                    Dim conversion As BoundExpression = _argument.InConversion
+                    ' The bound trees can contain a simple placeholder as a conversion. Don't treat this as a conversion.
+                    If conversion IsNot Nothing AndAlso conversion.Kind <> BoundKind.ByRefArgumentPlaceholder Then
```
Comment (genlu) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: Will the placeholder be rewritten into actual conversion later?

**Comment by genlu** [comment]
@JohnHamby I did a pass through IOperation changes, will take a closer look at tests later today.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb (hunk 6 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: No--in this case there is no actual conversion. My guess is that inserting the placeholders as "conversions" triggers generating code to do the copying in and out.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs [hunk 7]
```diff
@@ -0,0 +1,800 @@
...
+            Assert.Equal(argumentValue.Type.SpecialType, SpecialType.System_Double);
+            IConversionExpression conversion = (IConversionExpression)argumentValue;
+            argumentValue = ((IConversionExpression)argumentValue).Operand;
+            Assert.Equal(argumentValue.Type.SpecialType, SpecialType.System_Int32);
+            Assert.Equal(argumentValue.Kind, OperationKind.LocalReferenceExpression);
+            Assert.Equal(((ILocalReferenceExpression)argumentValue).Local.Name, "x");
+        }
+
+        [Fact]
+        public void ParamArrayInvocations()
```
Comment (genlu) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: Consider adding test for passing named argument to params array parameter, something like:
```cs
        public void M1(int first, params int[] more)
        { }

        public void TestNamedParamsArray(int first, params int[] more)
        {
            var x = new Class1();
            x.M1(more: new[] { 1, 2, 3 }, first: 0);
        }
```

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs [hunk 8]
```diff
@@ -0,0 +1,800 @@
...
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Semantics;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics
+{
+    public class IInvocationExpressionTests : CSharpTestBase
+    {
```
Comment (genlu) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: Are we interested in testing erroneous code for this exercise?

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs [hunk 9]
```diff
@@ -0,0 +1,800 @@
...
+            Assert.True(invocation.GetArgumentMatchingParameter(argument.Parameter) == argument);
+            argumentValue = argument.Value;
+            Assert.Equal(argumentValue.Kind, OperationKind.ConversionExpression);
+            Assert.False(argumentValue.IsInvalid);
+            Assert.Equal(argumentValue.Type.SpecialType, SpecialType.System_Double);
+            IConversionExpression conversion = (IConversionExpression)argumentValue;
+            argumentValue = ((IConversionExpression)argumentValue).Operand;
+            Assert.Equal(argumentValue.Type.SpecialType, SpecialType.System_Int32);
+            Assert.Equal(argumentValue.Kind, OperationKind.LocalReferenceExpression);
+            Assert.Equal(((ILocalReferenceExpression)argumentValue).Local.Name, "x");
```
Comment (genlu) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: It seems there's a pattern for many of those assertion groups, e.g. check invocation, check argument, etc. Consider creating and using helper methods if possible, probably would make the test easier to read.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs [hunk 10]
```diff
@@ -0,0 +1,800 @@
...
+            Assert.False(argumentValue.IsInvalid);
+            Assert.Equal(argumentValue.Type.SpecialType, SpecialType.System_Int32);
+            Assert.True(argumentValue.ConstantValue.HasValue);
+            Assert.Equal(argumentValue.ConstantValue.Value, 30);
+        }
+
+        [Fact]
+        public void DelegateInvocations()
+        {
+            const string source = @"
```
Comment (genlu) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: Delegate invocation might use named arguments too:
```cs
            System.Func<int, int, bool> f = (a, b) => a > b;
            f(arg1:1, arg2:2);
```

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs [hunk 11]
```diff
@@ -0,0 +1,800 @@
...
+        public void RefAndOutInvocations()
+        {
+            const string source = @"
+class C
+{
+    void M1()
+    {
+        int x = 10;
+        int y;
+        F(ref x, out y);
```
Comment (genlu) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: It seems this is also valid:
```cs
F(yy: out y, xx: ref x);
```

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 12]
```diff
@@ -98,52 +82,72 @@ internal static ImmutableArray<IArgument> DeriveArguments(ImmutableArray<BoundEx
                     argumentIndex = argumentsToParameters.IndexOf(parameterIndex);
                 }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Not part of this change but the assignment in the declaration of `argumentIndex` is unnecessary. Also consider using conditional so `argumentIndex` can be assigned in the declaration: ```int argumentIndex = argumentsToParameters.IsDefault ? parameterIndex : ...;```

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 13]
```diff
@@ -98,52 +82,72 @@ internal static ImmutableArray<IArgument> DeriveArguments(ImmutableArray<BoundEx
...
-                        // If the invocation is semantically valid, the parameter will be a params array and an empty array will be provided.
-                        // If the argument is otherwise omitted for a parameter with no default value, the invocation is not valid and a null argument will be provided.
-                        arguments.Add(DeriveArgument(parameterIndex, boundArguments.Length, boundArguments, argumentNames, argumentRefKinds, parameters, invocationSyntax));
-                    }
+            return arguments.ToImmutableAndFree();
+        }
+
+        internal static ImmutableArray<IArgument> DeriveArgumentsInEvaluationOrder(ImmutableArray<BoundExpression> boundArguments, ImmutableArray<string> argumentNames, ImmutableArray<int> argumentsToParameters, ImmutableArray<RefKind> argumentRefKinds, ImmutableArray<Symbols.ParameterSymbol> parameters, SyntaxNode invocationSyntax)
+        {
+            HashSet<int> matchedParameters = new HashSet<int>();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Would ```ArrayBuilder<bool>``` have less overhead than `HashSet<int>`, where the `bool` represents whether the parameter index was used?

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 14]
```diff
@@ -98,52 +82,72 @@ internal static ImmutableArray<IArgument> DeriveArguments(ImmutableArray<BoundEx
...
+            ArrayBuilder<IArgument> evaluationOrderArguments = ArrayBuilder<IArgument>.GetInstance(parameters.Length);
+            for (int argumentIndex = 0; argumentIndex < boundArguments.Length; argumentIndex++)
+            {
+                int parameterIndex = argumentsToParameters.IsDefault ? argumentIndex : argumentsToParameters[argumentIndex];
+                IArgument argument = DeriveArgument(parameterIndex, argumentIndex, boundArguments, argumentNames, argumentRefKinds, parameters, invocationSyntax);
+                evaluationOrderArguments.Add(argument);
+                matchedParameters.Add(parameterIndex);
+                if ((uint)parameterIndex < parameters.Length && parameters[parameterIndex].IsParams)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Is cast to `(uint)` necessary?

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 15]
```diff
@@ -45,26 +46,9 @@ internal partial class BoundCall : IInvocationExpression
...
-                        break;
-                    }
-                }
-
-                return sourceOrderArguments.ToImmutableAndFree();
-            }
-        }
-
-        ImmutableArray<IArgument> IHasArgumentsExpression.ArgumentsInParameterOrder => DeriveArguments(this.Arguments, this.ArgumentNamesOpt, this.ArgsToParamsOpt, this.ArgumentRefKindsOpt, this.Method.Parameters, this.Syntax);
+        ImmutableArray<IArgument> IHasArgumentsExpression.ArgumentsInEvaluationOrder => DeriveArgumentsInEvaluationOrder(this.Arguments, this.ArgumentNamesOpt, this.ArgsToParamsOpt, this.ArgumentRefKindsOpt, this.Method.Parameters, this.Syntax);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Does "evaluation order" mean order in source or order that the arguments are evaluated? (Are we providing guarantees on the evaluation order?)

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 16]
```diff
@@ -98,52 +82,72 @@ internal static ImmutableArray<IArgument> DeriveArguments(ImmutableArray<BoundEx
...
+            ArrayBuilder<IArgument> evaluationOrderArguments = ArrayBuilder<IArgument>.GetInstance(parameters.Length);
+            for (int argumentIndex = 0; argumentIndex < boundArguments.Length; argumentIndex++)
+            {
+                int parameterIndex = argumentsToParameters.IsDefault ? argumentIndex : argumentsToParameters[argumentIndex];
+                IArgument argument = DeriveArgument(parameterIndex, argumentIndex, boundArguments, argumentNames, argumentRefKinds, parameters, invocationSyntax);
+                evaluationOrderArguments.Add(argument);
+                matchedParameters.Add(parameterIndex);
+                if ((uint)parameterIndex < parameters.Length && parameters[parameterIndex].IsParams)
+                {
+                    break;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Is it correct to break here? What if named arguments are used and the named arguments following the `params` argument are not in parameter order?

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 17]
```diff
@@ -98,52 +82,72 @@ internal static ImmutableArray<IArgument> DeriveArguments(ImmutableArray<BoundEx
...
-            return arguments.ToImmutableAndFree();
+            return evaluationOrderArguments.ToImmutableAndFree();
         }
 
         private static readonly ConditionalWeakTable<BoundExpression, IArgument> s_argumentMappings = new ConditionalWeakTable<BoundExpression, IArgument>();
+        private static readonly ConditionalWeakTable<SyntaxNode, ConcurrentDictionary<Symbols.ParameterSymbol, IArgument>> s_omittedArgumentMappings = new ConditionalWeakTable<SyntaxNode, ConcurrentDictionary<Symbols.ParameterSymbol, IArgument>>();
 
         private static IArgument DeriveArgument(int parameterIndex, int argumentIndex, ImmutableArray<BoundExpression> boundArguments, ImmutableArray<string> argumentNames, ImmutableArray<RefKind> argumentRefKinds, ImmutableArray<Symbols.ParameterSymbol> parameters, SyntaxNode invocationSyntax)
         {
             if ((uint)argumentIndex >= (uint)boundArguments.Length)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Are casts to `(uint)` necessary?

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 18]
```diff
@@ -230,9 +237,9 @@ private static int ArgumentIndexMatchingParameter(ImmutableArray<BoundExpression
             return -1;
         }
 
-        private abstract class ArgumentBase : IArgument
+        private class Argument : IArgument
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: `sealed`?

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 15 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: My understanding that this should match the order of evaluation at runtime. For C#, it is the source order, if we exclude default parameters. For VB, it might not be the source order if arguments were named.  

---
In reply to: [64136328](https://github.com/dotnet/roslyn/pull/11450#discussion_r64136328) [](ancestors = 64136328)

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: I would suggest that this implementation be based on ArgumentsInEvaluationOrder. It should simply reorder arguments if needed. Otherwise, it should not derive arguments itself because there is a risk of that producing something different by comparison what ArgumentsInEvaluationOrder would produce for the same argument.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 19]
```diff
@@ -98,52 +82,72 @@ internal static ImmutableArray<IArgument> DeriveArguments(ImmutableArray<BoundEx
...
-                    else
-                    {
-                        // If the invocation is semantically valid, the parameter will be a params array and an empty array will be provided.
-                        // If the argument is otherwise omitted for a parameter with no default value, the invocation is not valid and a null argument will be provided.
-                        arguments.Add(DeriveArgument(parameterIndex, boundArguments.Length, boundArguments, argumentNames, argumentRefKinds, parameters, invocationSyntax));
-                    }
+            return arguments.ToImmutableAndFree();
+        }
+
+        internal static ImmutableArray<IArgument> DeriveArgumentsInEvaluationOrder(ImmutableArray<BoundExpression> boundArguments, ImmutableArray<string> argumentNames, ImmutableArray<int> argumentsToParameters, ImmutableArray<RefKind> argumentRefKinds, ImmutableArray<Symbols.ParameterSymbol> parameters, SyntaxNode invocationSyntax)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: This code is essentially doing what LocalRewriter does. There is a lot of logic duplicated here and there is a lot of opportunities to deviate from what LocalRewriter is doing. I believe our goal should be to return result consistent with what LocalRewriter would produce. As such we should find a way to share that logic, IOperation should not have its own private implementation.

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs [hunk 20]
```diff
@@ -529,14 +520,9 @@ public sealed override void Initialize(AnalysisContext context)
...
 
-                         if (argument.ArgumentKind == ArgumentKind.DefaultValue)
-                         {
-                             operationContext.ReportDiagnostic(Diagnostic.Create(UseDefaultArgumentDescriptor, invocation.Syntax.GetLocation(), argument.Parameter.Name));
-                         }
-
                          TestAscendingArgument(operationContext, argument.Value, ref priorArgumentValue);
 
-                         if (argument.ArgumentKind == ArgumentKind.ParamArray)
+                         if (argument.Parameter.IsParams)
```
Comment (AlekseyTs) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs]: Not sure about intent here. ParameterSymbol.IsParams doesn't guarantee that the method was used in expanded form.

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs [hunk 21]
```diff
@@ -1777,7 +1763,7 @@ public override void Visit(IOperation operation)
                     }
                     if (operation.Kind == OperationKind.Argument)
                     {
-                        if (((IArgument)operation).ArgumentKind == ArgumentKind.ParamArray)
+                        if (((IArgument)operation).Parameter.IsParams)
```
Comment (AlekseyTs) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs]: Not sure about intent here. ParameterSymbol.IsParams doesn't guarantee that the method was used in expanded form.

**Comment by AlekseyTs** [comment]
>                     ' Apparently the VB bound trees don't encode named arguments, which seems unnecesarily lossy.

Consider removing this comment as argument names in VB do not affect the order of evaluation.

---
Refers to: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb:375 in 4a94d83. [](commit_id = 4a94d834546817a81eaa858e970cd1d1fc6df461, deletion_comment = False)

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb [hunk 22]
```diff
@@ -379,9 +377,9 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                         argument,
                         Function(argumentValue)
                             If index >= parameters.Length - 1 AndAlso parameters.Length > 0 AndAlso parameters(parameters.Length - 1).IsParamArray Then
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: I do not understand why this conditional logic is needed.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs [hunk 23]
```diff
@@ -0,0 +1,800 @@
...
+        {
+            const string source = @"
+class C
+{
+    void M1()
+    {
+        M2(1, 2, 3);
+        M2(1);
+        M2();
+        M2(1, new int[] { 2, 3 });
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: Consider also testing ```M2(c: 1, a:2);```

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs (hunk 8 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: +1

---
In reply to: [64130139](https://github.com/dotnet/roslyn/pull/11450#discussion_r64130139) [](ancestors = 64130139)

**Comment by JohnHamby** [comment]
I will take care of fixing up TestOperationWalker.

**Comment by JohnHamby** [comment]
The commit just pushed is for purposes of backing up work in progress. The change set is not yet ready for further review.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs (hunk 7 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: Adding that test exposed a bug. I think the cause is reported elsewhere in this PR.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs (hunk 10 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: Added tests for named arguments in delegate invocation.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs (hunk 11 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: Added tests for named out and reference arguments.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 12 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: OK.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 14 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: The cast can be elided if and only if the argumentsToParameters array never contains any negative elements.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 16 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: It was not correct, for exactly the reason indicated. A test suggested by @genlu exposed the issue.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 17 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Yes, argumentIndex can be negative here.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 18 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Yes, this class should be sealed.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 2 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: The risk is low, but I will make ArgumentsInParameterOrder return ArgumentsInEvaluationOrder (reordered if necessary).

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 19 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Agreed. Unfortunately (for purposes of sharing) LocalRewriter does more than is desired in the way of transformations and requires information that is not easily available here. Packaging precisely the appropriate logic from LocalRewriter looks nontrivial to me.

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs (hunk 20 above)
Comment (JohnHamby) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs]: The intent is to analyzer arrays passed to params/ParamArray parameters in a uniform fashion whether the arguments were originally in array form or in expanded form.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 19 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: I think it is possible to find a way to share the code. Duplicating this complicated logic is not acceptable solution in the long run, I believe.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: I do not think there is any other way to be confident that the implementation here is correct.

File: src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs (hunk 21 above)
Comment (JohnHamby) [inline src/Compilers/Core/CodeAnalysisTest/Diagnostics/OperationTestAnalyzer.cs]: For purposes of this test, it doesn't matter whether or not the arguments were originally in array form or in expanded form.

**Comment by CyrusNajmabadi** [comment]
Why was ArgumentKind removed?  It seems *super* useful.

**Comment by JohnHamby** [comment]
ArgumentKind is removed because:
  1) It's a syntactic concept, not a semantic concept.
  2) The VB implementation would have been difficult to make correct, because the VB compiler does not presently preserve information like named arguments and paramarray arguments in the bound trees. Rummaging through the syntax trees to try to reconstruct the original form wasn't appealing.

**Comment by CyrusNajmabadi** [comment]
> It's a syntactic concept, not a semantic concept.

This is unfortunate.  Right now it's a bit hard for me to tell what the design around IOperation should be.  Right now it seems very oriented at exposing bound nodes.  And that's def a good thing.  However, at the same time it serves as a great way of providing a language neutral way of viewing C# and VB code.  

For many IDE features we'd like to use IOperation as a mechanism to allow us to write features without having to have duplicated logic for VB and C# to perform the same syntactic check.  IOperation has provided us with that neutral introspection API.  For example, PopulateSwitch is written without any C# of VB specific code at all: https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/PopulateSwitch/PopulateSwitchCodeFixProvider.cs

These two designs end up causing tension though.  There are things that VB and C# syntactically share which IOperation still does not necessarily represent.  While i see the benefit of just exposing the BoundNode data, i would still very much like it if we can push things further and allow this to be the language neutral way of operating over VB and C#.

So, for something like IArgument, having hte ArgumentKind is very nice.  It's true that it's a syntactic concept.  But, at the same time, it's common to C# and VB and otherwise requires duplicated code in a feature to ask the exact same for each language.

**Comment by JohnHamby** [comment]
Wow. I love the PopulateSwitch example, and would very much like to make more things like this possible.

Certainly I think ArgumentKind would be useful, else it wouldn't have been part of the original concept. If VB made supporting it easy I probably wouldn't be pushing to remove it.

Having a part of the API that is not supported correctly in VB is not tenable, and getting the VB implementation correct would require time that I don't have. (Anyone else want to take a whack at it?) The whole IOperation feature is under severe time pressure to meet correctness requirements. We can restore ArgumentKind to the API when there is time to implement it correctly in all supported languages.

**Comment by CyrusNajmabadi** [comment]
> The whole IOperation feature is under severe time pressure to meet correctness requirements. 

Is it an option to spend more time on this, and release as a more stable API later?  I'd rather spend more up front time on this than rush something through.

File: src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb (hunk 22 above)
Comment (JohnHamby) [inline src/Compilers/VisualBasic/Portable/BoundTree/Expression.vb]: This logic probably isn't needed. It's left over from before I discovered that the VB bound trees already include paramarray matching. I deleted the If statement and no tests failed.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs (hunk 23 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: That test finds a bug. The bug would probably be fixed by using the logic in LocalRewriter for params argument matching--I'll try to do that.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 13 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: A simple array is probably sufficient, given that the number of parameters is known in advance and the indices are contiguous. I'll look into this if this code survives.

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs (hunk 9 above)
Comment (JohnHamby) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/IInvocationExpressionTests.cs]: Yes, this turned out to be the case.

**Comment by JohnHamby** [comment]
The commit I just pushed will cause this PR to fail to build--I pushed the commit to save some work in progress.
